packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -3089,6 +3094,8 @@ export class ContainerRuntime
 					if (this.dirtyContainer !== checkpointDirtyState) {
 						this.updateDocumentDirtyState(checkpointDirtyState);
 					}
+					stageControls?.discardChanges();
```

Comment:
This will call `outbox.flush`, but without the if check that is below around the `this.flush` call.  (PS if we called `this.flush` here it would assert since the counter isn't back to 0 yet).

Reply:
it will only happen if the counter is zero as we only set the stage controls var when the counter is zero. all recursive calls will have undefined.